R/fcast.R

```diff
@@ -156,7 +156,7 @@ dcast.data.table = function(data, formula, fun.aggregate = NULL, sep = "_", ...,
   if (is.null(fun.call)) {
     oo = forderv(dat, by=varnames, retGrp=TRUE)
     if (attr(oo, 'maxgrpn', exact=TRUE) > 1L) {
-      messagef("Aggregate function missing, defaulting to 'length'")
+      messagef("Warning: Duplicated Combinations Detected. This occurs when combinations of rows and columns are not unique, meaning certain combinations representing rows and columns in the new data table repeat in the original data table, pointing to more than one value in original data table. However, we only require a single representative value for each combination in new data table. This can be achieved using an aggregation function, which consolidates multiple values into one as needed. Functions like mean or sum are suitable for this purpose. but If no aggregation function is explicitly specified by user, then the length of values (i.e., the number of different values associated with each combination) is used as the unique value. To customize this behavior, user can input their desired aggregation function to manipulate the values accordingly. For further details, type `?dcast` in the terminal.")
```

Comment:
This is a really great start and already a substantial improvement over the original message!

NB: Using markdown quotes (`>`) instead of a suggestion (```` ```suggestion ````) because the message is so long, I think it'll be easier to read than an inline diff of a very wide line.

Some high level comments about what changes I'm trying to make:

 1. Since this is a `message()`, "warning" is not the right framing of this output
 2. We should also make clear up front that this message is only coming because `fun.aggregate` was not provided
 3. "This occurs... more than one value in original data table." is a bit too wordy for me, see my attempt to rephrase and make sure it still conveys the idea to you (I also think we should make clear these rows/columns are derived from the `formula=` argument)
 4. We can also include `varnames` in the message to really hammer the point home

> 'fun.aggregate' is NULL, but found duplicate row/column combinations, so defaulting to length(). That is, the variables (%s) used in 'formula' do not uniquely identify rows in the input 'data'. In such cases, 'fun.aggregate' is used to derive a single representative value for each combination in the output data.table, for example by summing or averaging (fun.aggregate=sum or fun.aggregate=mean, respectively). See `?dcast.data.table` for more details.

Template value `%s` would be `brackify(varnames)`.

I'm still thinking of the right way to present insights about _what_ combinations are causing the multiplicity issue. The following is a very basic attempt to give something useful, but it suffers from not being robust enough

```r
print(dat[, .N, by=varnames][N > 1L])
```

The issue is that `varnames` _need not all be in `dat`_! Continuing the example from the issue comments (https://github.com/Rdatatable/data.table/issues/5217#issuecomment-1053960019)

```r
grp3 = DT$grp2
dcast(DT, grp1 ~ grp3, value.var="val")
```

Which complicates things a fair amount.